agent/acs/handler/attach_instance_eni_handler.go

```diff
@@ -0,0 +1,155 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package handler
+
+import (
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecsacs"
+	apieni "github.com/aws/amazon-ecs-agent/agent/api/eni"
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate"
+	"github.com/aws/amazon-ecs-agent/agent/statemanager"
+	"github.com/aws/amazon-ecs-agent/agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws"
+
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+
+	"context"
+)
+
+// attachInstanceENIHandler handles instance ENI attach operation for the ACS client
+type attachInstanceENIHandler struct {
+	messageBuffer     chan *ecsacs.AttachInstanceNetworkInterfacesMessage
+	ctx               context.Context
+	cancel            context.CancelFunc
+	saver             statemanager.Saver
+	cluster           *string
+	containerInstance *string
+	acsClient         wsclient.ClientServer
+	state             dockerstate.TaskEngineState
+}
+
+// newAttachInstanceENIHandler returns an instance of the attachInstanceENIHandler struct
+func newAttachInstanceENIHandler(ctx context.Context,
+	cluster string,
+	containerInstanceArn string,
+	acsClient wsclient.ClientServer,
+	taskEngineState dockerstate.TaskEngineState,
+	saver statemanager.Saver) attachInstanceENIHandler {
+
+	// Create a cancelable context from the parent context
+	derivedContext, cancel := context.WithCancel(ctx)
+	return attachInstanceENIHandler{
+		messageBuffer:     make(chan *ecsacs.AttachInstanceNetworkInterfacesMessage),
+		ctx:               derivedContext,
+		cancel:            cancel,
+		cluster:           aws.String(cluster),
+		containerInstance: aws.String(containerInstanceArn),
+		acsClient:         acsClient,
+		state:             taskEngineState,
+		saver:             saver,
+	}
+}
+
+// handlerFunc returns a function to enqueue requests onto attachENIHandler buffer
+func (handler *attachInstanceENIHandler) handlerFunc() func(message *ecsacs.AttachInstanceNetworkInterfacesMessage) {
+	return func(message *ecsacs.AttachInstanceNetworkInterfacesMessage) {
+		handler.messageBuffer <- message
+	}
+}
+
+// start invokes handleMessages to ack each enqueued request
+func (handler *attachInstanceENIHandler) start() {
+	go handler.handleMessages()
+}
+
+// stop is used to invoke a cancellation function
+func (handler *attachInstanceENIHandler) stop() {
+	handler.cancel()
+}
+
+// handleMessages handles each message one at a time
+func (handler *attachInstanceENIHandler) handleMessages() {
+	for {
+		select {
```

Comment:
You should also read `ctx.Done` channel here and exit the method if the context is done/cancelled. Not doing that leads to go routine leaks, which can cause problems in a long running agent.

Reply:
not sure if i understand.. i already listen to handler.ctx.Done in the following lines

Reply:
Sorry, I missed that. The convention is usually to have that as the first case statement. Probably why I missed that.

Having said that, even `sendAck` should handle the context being cancelled. Otherwise, that go routine can leak as well.

Reply:
Moved ctx.Done to first case statement. Regards to handling context in sendAck, I'm not sure how to do this since it just calls a function and returns rather than keeps doing something in a loop. Also the underlying vendor package gorilla/websocket that it uses does not handle context. Do you have a suggestion on how this can be done?

Reply:
I see. As far as `sendAck` itself is concerned, it's probably fine then. Note that we set read and write timeouts on the websocket channel itself to deal with `MakeRequest` taking too long.